src/Storages/MergeTree/tests/gtest_partition_action_blocker.cpp

```diff
@@ -0,0 +1,398 @@
+    PartitionActionBlocker blocker;
+
+    for (size_t i = 0; i < partitions_count; ++i)
+    {
+        blocker.cancelForPartition(partition_id + "_" + std::to_string(i));
+    }
+
+    // Automatic cleanup happens once in a while, 100 stale locks should trigger it.
+    EXPECT_LT(blocker.countPartitionBlockers(), partitions_count);
+}
+
+TEST(PartitionActionBlocker, TestCompactPartitionBlockers)
+{
+    const size_t partitions_count = 100;
+    const std::string partition_id = "some partition id";
+    PartitionActionBlocker blocker;
+
+    for (size_t i = 0; i < partitions_count; ++i)
+    {
+        blocker.cancelForPartition(partition_id + "_" + std::to_string(i));
+    }
+    // Manually cleanup all stale blockers (all blockers in this case).
+    blocker.compactPartitionBlockers();
+
+    EXPECT_EQ(0, blocker.countPartitionBlockers());
+}
+
+TEST(PartitionActionBlocker, TestCompactPartitionBlockersDoesntRemoveActiveBlockers)
+{
+    const size_t partitions_count = 100;
+    const std::string partition_id = "some partition id";
+    PartitionActionBlocker blocker;
+
+    auto lock_foo = blocker.cancelForPartition("FOO");
+    for (size_t i = 0; i < partitions_count; ++i)
+    {
+        blocker.cancelForPartition(partition_id + "_" + std::to_string(i));
+    }
+    auto lock_bar = blocker.cancelForPartition("BAR");
+
+    EXPECT_LT(2, blocker.countPartitionBlockers());
+
+    // Manually cleanup all stale blockers (all except held by lock_foo and lock_bar).
+    blocker.compactPartitionBlockers();
+
+    EXPECT_EQ(2, blocker.countPartitionBlockers());
+}
+
+TEST(PartitionActionBlocker, TestFormatDebug)
+{
+    // Do not validate contents, just make sure that something is printed out
+
+    const size_t partitions_count = 100;
+    const std::string partition_id = "some partition id";
+    PartitionActionBlocker blocker;
+
+    auto global_lock = blocker.cancel();
+    auto lock_foo = blocker.cancelForPartition("FOO");
+    auto lock_foo2 = blocker.cancelForPartition("FOO");
+    for (size_t i = 0; i < partitions_count; ++i)
+    {
+        blocker.cancelForPartition(partition_id + "_" + std::to_string(i));
+    }
+    auto lock_bar = blocker.cancelForPartition("BAR");
+
+    EXPECT_NE("", blocker.formatDebug());
+}
+
+template <>
+struct fmt::formatter<std::thread::id> : public fmt::formatter<std::string>
+{
+    auto format(const std::thread::id & id, fmt::format_context& ctx) const
+    {
+        std::stringstream sstr;
+        sstr << id;
+
+        return fmt::formatter<std::string>::format(sstr.str(), ctx);
+    }
+};
+
+template <typename R, typename P>
+struct fmt::formatter<std::chrono::duration<R, P>> : public fmt::formatter<std::string>
+{
+    auto format(const std::chrono::duration<R, P> & duration, fmt::format_context& ctx) const
+    {
+        std::stringstream sstr;
+        sstr << duration;
+
+        return fmt::formatter<std::string>::format(sstr.str(), ctx);
+    }
+};
+
+// TODO: remove this before merging
```

Comment:
@Enmk can this already be removed (alongside two definitions above)? Or is this still needed?